src/Chapter3.hs

```diff
@@ -480,6 +513,32 @@ After defining the city, implement the following functions:
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
 
+data OptionalWall = NoWall | Wall deriving (Show)
+data OptionalCastle = NoCastle | Castle String deriving (Show)
+data House = House { housePeople :: Int } deriving (Show)
+data PublicBuilding = Church | Library deriving (Show)
+data City = City
+  { cityCastle :: OptionalCastle
+  , cityWall :: OptionalWall
+  , cityBuilding :: PublicBuilding
+  , cityHouses :: [House]
+  } deriving (Show)
+
+buildCastle :: String -> City -> City
+buildCastle name city = city { cityCastle = (Castle name) }
+
+buildHouse :: Int -> City -> City
+buildHouse people city
+  | people >= 1 && people <= 4 = city { cityHouses = (House people) : (cityHouses city) }
```

Comment:
Interesting approach. It is usually called "smart constructor". To avoid building houses with more than 4 people you need to restrict usage of the Constructor and allow to create `House`s only with your `buildHouse` function.
You can read a bit more on the pattern in here: https://kowainik.github.io/posts/haskell-mini-patterns#smart-constructor

Generally, in here, you could restrict that on type level, by using something else than `Int` 🙂

Reply:
Oh, also spaces are redundant 👾

Reply:
Yes, I was considering having an enum for this but without knowing about existence of `Enum` typeclass calculating the 10-people limit would be ugly. Also I thought "would I use enum if houses could fit up to 40 persons?"